Fix BMW ConnectedDrive, update to My BMW API

rikroe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
- The following services/sensors have been removed due to not being available using the My BMW API: `alltrips_*`, `lasttrips_*`,  `chargingconnectiontype`, `charginginductivepositioning`, `lastchargingendreason`, `lastchargingendresult`, `maxelectricrange`
- `notify` requires a location attribute at `data.location`, as the MyBMW API only supports sending POI and not messages.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update `bimmer_connected` to `0.8.2` to adapt to new APIs by BMW. 
As these APIs return very different structures, a lot of changes were needed in the library and therefore some adjustments in the component as well.

The changes have been thoroughly tested via a custom component (see https://github.com/home-assistant/core/issues/59403).

Library release notes: 
- https://github.com/bimmerconnected/bimmer_connected/releases/tag/0.8.0
- https://github.com/bimmerconnected/bimmer_connected/releases/tag/0.8.1
- https://github.com/bimmerconnected/bimmer_connected/releases/tag/0.8.2

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #59403
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/20433

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rikroe committed:
Fix BMW ConnectedDrive, update to My BMW API

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -380,111 +133,10 @@ async def async_setup_entry(
+                        if (description := SENSOR_TYPES.get(attribute_name))
                     ]
                 )
-            if service == SERVICE_LAST_TRIP:
```
We can probably remove the pylint disable above since we remove a lot of nesting here.

rikroe replied:
Yes. Pylint succeeds when removed.

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -521,66 +173,45 @@ def __init__(
-                        date_str,
+            sensor_value = getattr(vehicle_state, sensor_key)
+
+            if isinstance(sensor_value, tuple):
```
Side note: It would be nice if we can define a data lookup and formatting function in the descriptions so we don't need all if/else checks here.

Example:
https://github.com/home-assistant/core/blob/59f10373ada8a784c2bd3cf0e8bfa91821baca80/homeassistant/components/airly/sensor.py#L120

rikroe replied:
Please check if this is the implementation you were thinking of!

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -521,66 +173,45 @@ def __init__(
+            charging_state = self._vehicle.status.charging_status in {
+                ChargingState.CHARGING
+            }
             self._attr_icon = icon_for_battery_level(
```
Side note: The battery device class sensor already sets icon depending on charge level so we should use device class battery and remove the use of the `icon_for_battery_level` function.

https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

Charging state should be shown by a binary sensor instead with device class `battery_charging`.

https://developers.home-assistant.io/docs/core/entity/binary-sensor#available-device-classes

rikroe replied:
Setting the icon attribute was removed. We already have a `charging_state` as binary sensor.

rikroe committed:
Use const device classes

rikroe committed:
Implement native_value via EntityDescription

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -53,6 +55,8 @@ class BMWSensorEntityDescription(SensorEntityDescription):
     "charging_status": BMWSensorEntityDescription(
         key="charging_status",
         icon="mdi:battery-charging",
+        device_class="battery",
```
Please import the sensor device class constants and use here.
```suggestion
        device_class=DEVICE_CLASS_BATTERY,
```

MartinHjelmare reviewed (commented):
Nice!

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -60,295 +55,75 @@ class BMWSensorEntityDescription(SensorEntityDescription):
+        unit_metric=PERCENTAGE,
+        unit_imperial=PERCENTAGE,
     ),
 }
```
We can remove the `DEFAULT_BMW_DESCRIPTION` now. It's not used.

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -519,68 +192,8 @@ def __init__(
         else:
             self._attr_native_unit_of_measurement = description.unit_metric
```
We can clean up some parameters and attributes now from the entity init. `service` parameter and `self._service` attribute aren't used anymore.

rikroe replied:
They might be coming back in one of the future versions of the library. If needed I can remove them though.

MartinHjelmare replied:
Yes, please remove code we don't use for now.

rikroe committed:
Use device class const, reomve device_class from charging_status

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -60,295 +55,75 @@ class BMWSensorEntityDescription(SensorEntityDescription):
     "charging_status": BMWSensorEntityDescription(
         key="charging_status",
```
Should this be `"charging_level_hv"` instead?

MartinHjelmare replied:
Nevermind I see we use that key below. What does the charging status value show? Is that an enum?

rikroe replied:
Yes, it is an additional info to the binary sensor for charging. https://github.com/bimmerconnected/bimmer_connected/blob/dev/bimmer_connected/vehicle_status.py#L38-L47

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -59,296 +55,76 @@ class BMWSensorEntityDescription(SensorEntityDescription):
+        icon="mdi:ev-station",
+        value=lambda x, y: x.value,
     ),
     # No icon as this is dealt with directly as a special case in icon()
```
```suggestion
```

rikroe committed:
Cleanup

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -174,20 +162,11 @@ def __init__(
-            self._attr_name = f"{vehicle.name} {service.lower()}_{description.key}"
-            self._attr_unique_id = f"{vehicle.vin}-{service.lower()}-{description.key}"
-        else:
-            self._attr_name = f"{vehicle.name} {description.key}"
```
We still want to set name and unique_id attributes.

rikroe replied:
Oops. I have re-added them in https://github.com/home-assistant/core/pull/59881/commits/ca7676ed97dadba099487d3bdba9071bd58cdfa2. However it doesn't seem to be reflected in the PR?!?

rikroe committed:
Remove max_range_electric

rikroe committed:
Revert removing sensor name & unique_id

rikroe committed:
Add region china again, update bimmer_connected

rikroe commented:
I have updated the main PR comment, as `bimmer_connected==0.8.1` now again supports `china`.

Removed breaking changes:
- ~Support for region `china` dropped (see https://github.com/bimmerconnected/bimmer_connected/issues/347 for details & workaround).~

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/manifest.json:
```diff
@@ -2,7 +2,7 @@
   "name": "BMW Connected Drive",
   "documentation": "https://www.home-assistant.io/integrations/bmw_connected_drive",
-  "requirements": ["bimmer_connected==0.7.21"],
+  "requirements": ["bimmer_connected==0.8.1"],
```
In 0.8.1 `pycryptodome` was added as a requirement and pinned to the exact version 3.11.0. We want to avoid that libraries pin requirements to an exact version to make the libraries more compatible with other libraries in Home Assistant that have the same requirements. Please loosen the pinning and require only a minimum version or a range with as forgiving limits as possible.

Thanks!

rikroe replied:
Done. Of course that makes a lot of sense!

rikroe replied:
Oops, test fail as `bimmer_connected==0.8.2` cannot be downloaded. But it has been released to PyPi: https://pypi.org/project/bimmer-connected/#history

MartinHjelmare replied:
I restarted CI.

rikroe committed:
Update to bimmer_connected==0.8.2

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/notify.py:
```diff
@@ -84,6 +81,4 @@ def send_message(self, message: str = "", **kwargs: Any) -> None:
 
                 vehicle.remote_services.trigger_send_poi(location_dict)
             else:
```
A future improvement: Invert the if check above and raise the ValueError if True. Then we can outdent below.

rikroe replied:
Yes, that makes sense. Will do that when your larger change to services/notify is merged so we have to update it only once.

MartinHjelmare commented:
Do we need a docs PR?

rikroe commented:
Yes, as we removed the messages. PR just created: https://github.com/home-assistant/home-assistant.io/pull/20433

rikroe commented:
Many thanks for your help and patience!

MartinHjelmare commented:
We don't seem to mention the removal of the notify messages. Please update the breaking change paragraph if needed.

rikroe commented:
Enhanced the second sentence in breaking changes to make it more clear.
